ftl/python/builder.py

```diff
@@ -0,0 +1,148 @@
+                                 % checksum)
+            else:
+                logging.info('No cached dependency layer for %s' % checksum)
+        else:
+            logging.info(
+                'Skipping checking cache for dependency layer %s' % checksum)
+
+        layer, sha = self._gen_package_tar(descriptor, descriptor_contents)
+
+        with append.Layer(
+                base_image, layer, diff_id=sha,
+                overrides=overrides) as dep_image:
+            if use_cache:
+                logging.info('Storing layer %s in cache.', sha)
+                cache.Store(base_image, _PYTHON_NAMESPACE, checksum, dep_image)
+            else:
+                logging.info('Skipping storing layer %s in cache.', sha)
+            return dep_image
+
+    def _gen_package_tar(self, descriptor, descriptor_contents):
+        tmp_app = tempfile.mkdtemp()
+        tmp_venv = tempfile.mkdtemp()
+
+        tmp_app = os.path.join(tmp_app, 'app')
+        venv_dir = os.path.join(tmp_venv, 'env')
+        os.makedirs(tmp_app)
+        os.makedirs(venv_dir)
+
+        # Copy out the relevant package descriptors to a tempdir.
+        with open(os.path.join(tmp_app, descriptor), 'w') as f:
+            f.write(descriptor_contents)
+
+        tar_path = tempfile.mktemp()
+        logging.info('Starting venv creation ...')
+
+        # TODO(aaron-prindle) add support for different python versions
+        subprocess.check_call(
+            ['virtualenv', '--no-download', venv_dir, '-p', 'python3.6'],
+            cwd=tmp_app)
+        os.environ['VIRTUAL_ENV'] = venv_dir
+        os.environ['PATH'] = venv_dir + "/bin" + ":" + os.environ['PATH']
+        # bazel adds its own PYTHONPATH to the env
+        # which must be removed for the pip calls to work properly
+        my_env = os.environ.copy()
+        my_env.pop('PYTHONPATH', None)
+
+        subprocess.check_call(
+            ['pip', 'install', '-r', 'requirements.txt'],
+            cwd=tmp_app,
+            env=my_env)
+        logging.info('Finished pip install.')
+
+        logging.info('Starting to tar pip packages...')
+        subprocess.check_call(['tar', '-C', tmp_venv, '-cf', tar_path, '.'])
+        logging.info('Finished generating tarfile for pip packages...')
+
+        # We need the sha of the unzipped and zipped tarball.
+        # So for performance, tar, sha, zip, sha.
+        # We use gzip for performance instead of python's zip.
+        sha = 'sha256:' + hashlib.sha256(open(tar_path).read()).hexdigest()
+
+        logging.info('Starting to gzip pip package tarfile...')
+        subprocess.check_call(['gzip', tar_path])
+        logging.info('Finished generating gzip pip package tarfile.')
+        return open(os.path.join(tmp_venv, tar_path + '.gz'), 'rb').read(), sha
+
+
+def From(ctx):
+    return Python(ctx)
+
+
+def _creation_time(image):
+    logging.info(image.config_file())
+    cfg = json.loads(image.config_file())
+    return cfg['created']
```

Comment:
Can you use a .get here in case there's no cfg['created']

Reply:
done